packages/bridge-ui-v2/src/libs/bridge/Prover.ts

```diff
@@ -0,0 +1,166 @@
+  }
+
+  private static async _getLatestBlock(
+    crossChainSyncContract: GetContractResult<typeof crossChainSyncABI>,
+    provider: PublicClient,
+  ) {
+    const latestBlockHash = await crossChainSyncContract.read.getCrossChainBlockHash([BigInt(0)]);
+    return provider.getBlock({ blockHash: latestBlockHash });
+  }
+
+  private static _getSignalProof(proof: EthGetProofResponse, blockHeight: bigint) {
+    // RLP encode the proof together for LibTrieProof to decode
+    const encodedProof = toRlp(proof.storageProof[0].proof);
+
+    // Encode the SignalProof struct:
+    // struct SignalProof {
+    //   uint256 height;
+    //   bytes proof;
+    // }
+    const signalProof = encodeAbiParameters(
+      // ['tuple(uint256 height, bytes proof)'],
+      [
+        {
+          type: 'tuple',
+          components: [
+            { name: 'height', type: 'uint256' },
+            { name: 'proof', type: 'bytes' },
+          ],
+        },
+      ],
+      [{ height: blockHeight, proof: encodedProof }],
+    );
+
+    return signalProof;
+  }
+
+  async generateProofToClaim({
+    msgHash,
+    chainId,
+    sender,
+    srcChainId,
+    destChainId,
+    destCrossChainSyncAddress,
+    srcSignalServiceAddress,
+  }: GenerateProofClaimArgs) {
+    // Get the block from the source chain based on the latest block hash
+    // we get cross chain (Taiko contract on the destination chain)
+
+    const crossChainSyncContract = getContract({
+      chainId: destChainId,
+      address: destCrossChainSyncAddress,
+      abi: crossChainSyncABI,
+    });
+
+    const provider = publicClient({ chainId: srcChainId });
+    const block = await Prover._getLatestBlock(crossChainSyncContract, provider);
+
+    if (block.hash === null || block.number === null) {
+      throw new PendingBlockError('block is pending');
+    }
+
+    const key = await Prover._getKey(sender, msgHash);
+
+    // Unfortunately, since this method is stagnant, it hasn't been included into Viem lib
+    // as supported methods. Still stupported  by Alchmey, Infura and others.
+    // See https://eips.ethereum.org/EIPS/eip-1186
+    // Following is a workaround to support this method.
+    const clientWithEthProofRequest = publicClient({ chainId }) as ClientWithEthProofRequest;
+
+    // RPC call to get the merkle proof what value is at key on the SignalService contract
+    const proof = await clientWithEthProofRequest.request({
+      method: 'eth_getProof',
+      params: [
+        // Address of the account to get the proof for
+        srcSignalServiceAddress,
+
+        // Array of storage-keys that should be proofed and included
+        [key],
+
+        block.hash,
+      ],
+    });
+
+    log('Proof from eth_getProof', proof);
+
+    // Value must be 1 => isSignalSent
+    if (proof.storageProof[0].value !== '0x1') {
+      throw new InvalidProofError('storage proof value is not 1: SignalSent');
+    }
```

Comment:
I changed the condition to make sure we've received the signal to be able to claim as 
```ts
proof.storageProof[0].value !== toHex(true)
```
and the condition to make sure the transaction has failed to be able to release as
```ts
proof.storageProof[0].value !== toHex(MessageStatus.FAILED)
```
which makes it clearer.

Regarding having a reusable Prover, this is a proof service exclusively for the bridge. Maybe the naming isn't clear? would it then make sense to call it ProofService instead?

Reply:
No, the old bridge doesn't have this design... Let me think about this. The thing is, we don't even need a class anymore here, since there is no state at all. I'm gonna change this

Reply:
https://github.com/taikoxyz/taiko-mono/blob/0da63995cf94f463eece3fdd3a70cfb9d08ad395/packages/bridge-ui/src/proof/ProofService.ts#L16... Not sure where this abstraction you mentioned is here. The ProofService is implementing an interface. But that's about it. There is no need for any class here. If we want something more general, I can separate the proof generation and the one needed to claim, which will throw if there is no 1 stored in the signal service when we want to claim, and failed value in the destination bridge contract when we want to release. You know what I mean?

Reply:
Alright, was just wondering in case I was missing something here. The copy paste from previous version still has the issue you mentioned initially:
```ts
if (proof.storageProof[0].value !== '0x1') {
      throw new InvalidProofError('storage proof value is not 1: SignalSent');
    }
```
The only difference from this and the original one is the lack of interface, which it's not needed, small mods and comments in order to explain a bit more what's going on here.

Feel free to jump out then